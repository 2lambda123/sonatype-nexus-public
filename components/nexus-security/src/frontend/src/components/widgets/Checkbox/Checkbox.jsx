/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import PropTypes from 'prop-types';
import React from 'react';

import './Checkbox.scss';

export default function Checkbox({name, isChecked, onChange, labelText}) {
  return <label>
    <input
        name={name}
        type='checkbox'
        checked={isChecked}
        onChange={onChange}
        className='nxrm-checkbox'
    />
    {labelText}
  </label>;
}

Checkbox.propTypes = {
  name: PropTypes.string,
  isChecked: PropTypes.bool.isRequired,
  onChange: PropTypes.func,
  labelText: PropTypes.string
};
