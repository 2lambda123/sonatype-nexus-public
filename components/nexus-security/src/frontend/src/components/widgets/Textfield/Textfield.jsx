/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import './Textfield.scss';

import RequiredErrorMessage from '../RequiredErrorMessage/RequiredErrorMessage';

export default function Textfield({name, value, onChange, isRequired, className}) {
  const isMissingRequiredValue = isRequired && !value;
  const classes = classNames('nxrm-textfield', className, {
    'missing-required-value': isMissingRequiredValue
  });
  return <>
    <input
        name={name}
        type='text'
        value={value}
        onChange={onChange}
        className={classes}
    />
    {isMissingRequiredValue ? <RequiredErrorMessage/> : null}
  </>;
}

Textfield.propTypes = {
  name: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func,
  isRequired: PropTypes.bool
};
