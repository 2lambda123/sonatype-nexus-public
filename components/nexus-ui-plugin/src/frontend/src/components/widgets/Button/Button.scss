/**
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */

@import '../../../styles/colors';

.nx-btn {
  background-color: $almost-white;
  border: solid 1px $almost-black;
  border-radius: 3px;
  color: $almost-black;
  font-family: inherit;
  font-size: 1em;
  height: 2em;
  min-width: 76px;
  opacity: 1;
  cursor: pointer;

  &:hover:not(:disabled) {
    filter: brightness(0.9);
  }

  &:disabled {
    cursor: default;
    opacity: 0.5;
  }

  &:focus, &:active {
    border-color: $light-blue;
  }

  &.nx-btn--primary {
    background-color: $blue;
    color: $white;
  }
}
