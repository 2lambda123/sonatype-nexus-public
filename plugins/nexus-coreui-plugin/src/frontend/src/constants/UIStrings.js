/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import { UIStrings } from 'nexus-ui-plugin';
import React from 'react';

export default {
  ...UIStrings,

  ANONYMOUS_SETTINGS: {
    MENU: {
      text: 'Anonymous Access',
      description: 'Configure anonymous access to server contents'
    },
    ENABLED_CHECKBOX_LABEL: 'Access:',
    ENABLED_CHECKBOX_DESCRIPTION: 'Allow anonymous users to access the server',
    USERNAME_TEXTFIELD_LABEL: 'Username:',
    REALM_SELECT_LABEL: 'Realm:',
    MESSAGES: {
      LOAD_ERROR: 'An error occurred while loading Anonymous settings, see console for more details',
      SAVE_SUCCESS: 'Anonymous security settings updated',
      SAVE_ERROR: 'An error occurred while updating Anonymous settings, see console for more details'
    }
  },

  SYSTEM_INFORMATION: {
    MENU: {
      text: 'System Information'
    },
    ACTIONS: {
      download: 'Download as JSON'
    }
  },

  SUPPORT_REQUEST: {
    MENU: {
      text: 'Support Request',
      description: 'Submit a support request to Sonatype'
    },
    ACTIONS: {
      submitRequest: 'Submit Request'
    },
    DESCRIPTION: `Please include a complete description of your problem and steps to allow us to reproduce the problem (if available).`,
    ATTACH_SUPPORT_ZIP: `Attaching a support ZIP to your request will help our engineers give you a faster response.`
  },

  USER_ACCOUNT: {
    MENU: {
      text: 'Account',
      description: 'Manage your account',
    },
    ACTIONS: {
      changePassword: 'Change password',
      discardChangePassword: 'Discard password change',
    },
    MESSAGES: {
      LOAD_ERROR: 'An error occurred while loading User Account, see console for more details',
      UPDATE_SUCCESS: 'User account settings updated',
      UPDATE_ERROR: 'An error occurred while updating user account settings',
      PASSWORD_NO_MATCH_ERROR: 'Passwords do not match',
      PASSWORD_MUST_DIFFER_ERROR: 'New password must be different',
    },
    ID_FIELD_LABEL: 'Username',
    FIRST_FIELD_LABEL: 'First Name',
    LAST_FIELD_LABEL: 'Last Name',
    EMAIL_FIELD_LABEL: 'Email',
    PASSWORD_CURRENT_FIELD_LABEL: 'Current Password',
    PASSWORD_NEW_FIELD_LABEL: 'New Password',
    PASSWORD_NEW_CONFIRM_FIELD_LABEL: 'New Password (Confirm)',
  },
};
