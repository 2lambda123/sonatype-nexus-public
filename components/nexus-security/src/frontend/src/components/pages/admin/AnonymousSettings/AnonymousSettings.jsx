/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React, {useState, useEffect} from 'react';

import './AnonymousSettings.scss';

import Axios from 'axios';
import Button from '../../../../components/widgets/Button/Button';
import Checkbox from '../../../../components/widgets/Checkbox/Checkbox';
import ContentBody from '../../../../components/layout/common/ContentBody/ContentBody';
import ExtJS from '../../../../interface/ExtJS';
import Select from '../../../../components/widgets/Select/Select';
import SettingsSection from '../../../../components/layout/admin/SettingsSection/SettingsSection';
import Textfield from '../../../../components/widgets/Textfield/Textfield';
import UIStrings from '../../../../constants/UIStrings';

export default function AnonymousSettings() {
  const [anonymousSettings, setAnonymousSettings] = useState({
    enabled: false,
    realmName: '',
    userId: ''
  });
  const [pristineSettings, setPristineSettings] = useState({});
  const [realmTypes, setRealmTypes] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (!isLoading) {
      return;
    }

    const cancelTokenSource = Axios.CancelToken.source();
    const config = {
      cancelToken: cancelTokenSource.token
    };

    Axios.all([
      Axios.get('/service/rest/internal/ui/realms/types', config),
      Axios.get('/service/rest/internal/ui/anonymous-settings', config)
    ])
        .then(Axios.spread((realmTypes, settings) => {
          setIsLoading(false);
          setRealmTypes(realmTypes.data);
          setAnonymousSettings({...settings.data});
          setPristineSettings({...settings.data});
        }))
        .catch((error) => {
          if (!Axios.isCancel(error)) {
            ExtJS.showErrorMessage(UIStrings.ANONYMOUS_SETTINGS.MESSAGES.LOAD_ERROR);
            console.error(error);
          }
        });

    return function cleanup() {
      cancelTokenSource.cancel();
    }
  });

  const handleInputChange = (event) => {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    setAnonymousSettings({
      ...anonymousSettings,
      [name]: value
    });
  };

  const handleDiscard = () => setAnonymousSettings({...pristineSettings});

  const handleSave = () => {
    Axios.put('/service/rest/internal/ui/anonymous-settings', anonymousSettings)
        .then(() => {
          setPristineSettings({...anonymousSettings});
          ExtJS.showSuccessMessage(UIStrings.ANONYMOUS_SETTINGS.MESSAGES.SAVE_SUCCESS);
        })
        .catch((error) => {
          ExtJS.showErrorMessage(UIStrings.ANONYMOUS_SETTINGS.MESSAGES.SAVE_ERROR);
          console.error(error);
        });
  };

  const isPristine = Object.keys(pristineSettings).every(
      (key) => pristineSettings[key] === anonymousSettings[key]
  );
  const {enabled, userId, realmName} = anonymousSettings;

  return <ContentBody className='nxrm-anonymous-settings'>
    <SettingsSection isLoading={isLoading}>
      <SettingsSection.FieldWrapper
          labelText={UIStrings.ANONYMOUS_SETTINGS.ENABLED_CHECKBOX_LABEL}
      >
        <Checkbox
            name='enabled'
            isChecked={enabled}
            onChange={handleInputChange}
            labelText={UIStrings.ANONYMOUS_SETTINGS.ENABLED_CHECKBOX_DESCRIPTION}
        />
      </SettingsSection.FieldWrapper>
      <SettingsSection.FieldWrapper
          labelText={UIStrings.ANONYMOUS_SETTINGS.USERNAME_TEXTFIELD_LABEL}
      >
        <Textfield
            name='userId'
            value={userId}
            onChange={handleInputChange}
            isRequired={!isLoading}
            className='nxrm-anonymous-settings-field'
        />
      </SettingsSection.FieldWrapper>
      <SettingsSection.FieldWrapper
          labelText={UIStrings.ANONYMOUS_SETTINGS.REALM_SELECT_LABEL}
      >
        <Select
            name='realmName'
            value={realmName}
            onChange={handleInputChange}
            className='nxrm-anonymous-settings-field'
        >
          {
            realmTypes.map((realmType) =>
                <option key={realmType.id} value={realmType.id}>{realmType.name}</option>
            )
          }
        </Select>
      </SettingsSection.FieldWrapper>
      <SettingsSection.Footer>
        <Button
            isPrimary={true}
            disabled={isPristine || !userId}
            onClick={handleSave}
        >
          {UIStrings.SETTINGS.SAVE_BUTTON_LABEL}
        </Button>
        <Button
            disabled={isPristine}
            onClick={handleDiscard}
        >
          {UIStrings.SETTINGS.DISCARD_BUTTON_LABEL}
        </Button>
      </SettingsSection.Footer>
    </SettingsSection>
  </ContentBody>;
}
