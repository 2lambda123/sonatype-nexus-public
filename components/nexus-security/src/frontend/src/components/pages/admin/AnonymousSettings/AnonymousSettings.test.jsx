/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import {mount, shallow} from 'enzyme';
import React from 'react';

import Axios from 'axios';
import AnonymousSettings from './AnonymousSettings';
import Button from '../../../../components/widgets/Button/Button';
import Checkbox from '../../../../components/widgets/Checkbox/Checkbox';
import Select from '../../../../components/widgets/Select/Select';
import Textfield from '../../../../components/widgets/Textfield/Textfield';
import UIStrings from '../../../../constants/UIStrings';

const mockRealmTypes = [
  {id: 'r1', name: 'Realm One'},
  {id: 'r2', name: 'Realm Two'}
];
const mockAnonymousSettings = {
  enabled: true,
  userId: 'testUser',
  realmName: 'r2'
};

jest.mock('axios', () => {  // Mock out parts of axios, has to be done in same scope as import statements
  return {
    ...jest.requireActual('axios'), // Use most functions from actual axios
    get: jest.fn((url) => {
      switch (url) {
        case '/service/rest/internal/ui/realms/types':
          return Promise.resolve({data: mockRealmTypes});
        case '/service/rest/internal/ui/anonymous-settings':
          return Promise.resolve({data: mockAnonymousSettings});
      }
    }),
    put: jest.fn(() => Promise.resolve())
  };
});

jest.mock('../../../../interface/ExtJS');

describe('AnonymousSettings', () => {
  // see https://github.com/airbnb/enzyme/issues/1587
  const waitForDataFromApi = (wrapper) => Promise.resolve(wrapper)
      .then(() => wrapper.update())
      .then(() => wrapper.update());

  const assertDefaultValues = (wrapper) => {
    expect(wrapper.find(Checkbox)).toHaveProp('isChecked', mockAnonymousSettings.enabled);
    expect(wrapper.find(Textfield)).toHaveProp('value', mockAnonymousSettings.userId);
    expect(wrapper.find(Select)).toHaveProp('value', mockAnonymousSettings.realmName);
  };

  it('renders correctly', () => {
    expect(shallow(<AnonymousSettings/>)).toMatchSnapshot();
  });

  it('fetches the values of fields from the API and updates them as expected', async () => {
    // using mount because of bug preventing useEffect hook from being called when component is shallow rendered
    // See https://github.com/airbnb/enzyme/issues/2086
    const wrapper = mount(<AnonymousSettings/>);

    await waitForDataFromApi(wrapper);

    expect(Axios.get).toHaveBeenCalledTimes(2);
    assertDefaultValues(wrapper);
  });

  it('Sends changes to the API on save', async () => {
    // using mount because of bug preventing useEffect hook from being called when component is shallow rendered
    // See https://github.com/airbnb/enzyme/issues/2086
    const wrapper = mount(<AnonymousSettings/>);

    await waitForDataFromApi(wrapper);

    expect(Axios.get).toHaveBeenCalledTimes(2);
    expect(Axios.put).toHaveBeenCalledTimes(0);

    wrapper.find(Textfield).simulate('change', {
      target: {
        type: 'text',
        name: 'userId',
        value: 'changed-username'
      }
    });

    wrapper.findWhere(node =>
        node.is(Button) && node.text() === UIStrings.SETTINGS.SAVE_BUTTON_LABEL
    ).simulate('click');

    expect(Axios.put).toHaveBeenCalledTimes(1);
    expect(Axios.put).toHaveBeenCalledWith(
        '/service/rest/internal/ui/anonymous-settings',
        {
          ...mockAnonymousSettings,
          userId: 'changed-username'
        }
    );
  });

  it('Sends nothing to the API on discard', async () => {
    // using mount because of bug preventing useEffect hook from being called when component is shallow rendered
    // See https://github.com/airbnb/enzyme/issues/2086
    const wrapper = mount(<AnonymousSettings/>);

    await waitForDataFromApi(wrapper);

    wrapper.find(Select).simulate('change', {
      target: {
        name: 'realmName',
        value: 'r2'
      }
    });

    wrapper.findWhere(node =>
        node.is(Button) && node.text() === UIStrings.SETTINGS.DISCARD_BUTTON_LABEL
    ).simulate('click');

    expect(Axios.put).toHaveBeenCalledTimes(0);
    assertDefaultValues(wrapper);
  });

  it('Does not allow the user to save if the required username is empty', async () => {
    // using mount because of bug preventing useEffect hook from being called when component is shallow rendered
    // See https://github.com/airbnb/enzyme/issues/2086
    const wrapper = mount(<AnonymousSettings/>);

    await waitForDataFromApi(wrapper);

    wrapper.find(Select).simulate('change', {
      target: {
        name: 'userId',
        value: ''
      }
    });

    const saveButton = wrapper.findWhere(node =>
        node.is(Button) && node.text() === UIStrings.SETTINGS.SAVE_BUTTON_LABEL
    );

    expect(saveButton).toBeDisabled();
  });
});
