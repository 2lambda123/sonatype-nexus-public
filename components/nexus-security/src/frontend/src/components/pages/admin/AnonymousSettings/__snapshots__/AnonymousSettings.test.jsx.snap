// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnonymousSettings renders correctly 1`] = `
<ContentBody
  className="nxrm-anonymous-settings"
>
  <SettingsSection
    isLoading={true}
  >
    <Component
      labelText="Access:"
    >
      <Checkbox
        isChecked={false}
        labelText="Allow anonymous users to access the server"
        name="enabled"
        onChange={[Function]}
      />
    </Component>
    <Component
      labelText="Username:"
    >
      <Textfield
        className="nxrm-anonymous-settings-field"
        isRequired={false}
        name="userId"
        onChange={[Function]}
        value=""
      />
    </Component>
    <Component
      labelText="Realm:"
    >
      <Select
        className="nxrm-anonymous-settings-field"
        name="realmName"
        onChange={[Function]}
        value=""
      />
    </Component>
    <Component>
      <Button
        disabled={true}
        isPrimary={true}
        onClick={[Function]}
      >
        Save
      </Button>
      <Button
        disabled={true}
        onClick={[Function]}
      >
        Discard
      </Button>
    </Component>
  </SettingsSection>
</ContentBody>
`;
